localstack/services/s3/presigned_url.py

```diff
@@ -597,29 +468,189 @@ def validate_presigned_url_s3v4(context: RequestContext) -> None:
+        self._set_aws_request(request_url=self.request_url)
+
+    @property
+    def request_url(self) -> str:
+        return f"{self.request.scheme}://{self.host}{self.path}?{self.request_query_string}"
+
+    def get_signature_data(self) -> Tuple[str, str, str]:
+        """
+        Uses the signer to return the signature and the data used to calculate it
+        :return: signature, canonical_request and string_to_sign
+        """
+        return self.signer.add_auth(self.aws_request, self.signature_date)  # noqa
+
+    def _get_signed_headers_and_filtered_query_string(self) -> Tuple[Dict[str, str], str]:
+        """
+        Transforms the original headers and query parameters to the headers and query string used to sign the
+        original request.
+        Allows us to recreate the original request.
+        :raises AccessDenied if the request contains headers that were not in X-Amz-SignedHeaders and started with x-amz
+        :return: the headers used to sign the request and the query string without X-Amz-Signature
+        """
+        headers = copy.copy(self._headers)
+        # set automatically by the handler chain, we don't want that
+        headers.pop("Authorization", None)
+        signed_headers = self._query_parameters.get("X-Amz-SignedHeaders")
+
+        signature_headers = {}
+        not_signed_headers = []
+        for header, value in headers.items():
+            header_low = header.lower()
+            if header_low.startswith("x-amz-"):
+                if header_low in IGNORED_SIGV4_HEADERS:
+                    continue
+                if header_low not in signed_headers.lower():
+                    not_signed_headers.append(header_low)
+            if header_low in signed_headers:
+                signature_headers[header_low] = value
+
+        if not_signed_headers:
+            ex: AccessDenied = create_access_denied_headers_not_signed(
+                ", ".join(not_signed_headers)
+            )
+            raise ex
+
+        new_query_args = {
+            arg: value for arg, value in self._query_parameters.items() if arg != "X-Amz-Signature"
+        }
+        new_query_string = percent_encode_sequence(new_query_args)
+        return signature_headers, new_query_string
+
+    def _set_aws_request(self, request_url: str) -> None:
+        """
+        Creates and sets the AWSRequest needed for S3SigV4QueryAuth signer
+        :param request_url: the request_url used for the calculation
+        :return:
+        """
+        request_dict = {
+            "method": self._request_method,
+            "url": request_url,
+            "body": b"",
+            "headers": self.signed_headers,
+            "context": {
+                "is_presign_request": True,
+                "use_global_endpoint": True,
+                "signing": {"bucket": self._bucket},
+            },
+        }
+        self.aws_request: AWSRequest = create_request_object(request_dict)
```

Comment:
nit: maybe it would be good to add the annotation to the class declaration.

Reply:
refactored to return the `AWSRequest` from the method 👍